### ISSUE TITLE
Noop specification deserves clarification(s)

### ISSUE THREAD

**Issue by DbxDev** [ubjson/universal-binary-json#80]
Noop is defined as a valueless value that can be added to the elements of a container.
That is not always true, in particular when the container is strongly typed and the type marker is omitted. In this case, there is no way to distinguish [N] from the integer 78.
For this particular reason, using noop as a way to remove data should be used with care, and is in practice not efficient because you first need to find how the container is defined (with optimized format or not).

**Comment by rkalla** [comment]
Definitely a miss on my part, that's a great point.

**Comment by tsieprawski** [comment]
@DbxDev, what to you mean by `when the container is strongly typed and the type marker is omitted. In this case, there is no way to distinguish [N] from the integer 78`? According to [1] example, if type marker is ommited, then container is not optimized by type at all. And if type marker's argument is absent, then probably parser will fail. And if both type marker and argument `N` are present, then every value in container is noop already.

Do I think correctly?

[1] https://github.com/ubjson/universal-binary-json/blob/master/spec12/container-types.html#L146

**Comment by rkalla** [comment]
You are correct - the only way to omit the type marker for the container is
if all the entries are typed... if you want to omit the type marker on the
entries, then the container (value) must be typed.

**Comment by DbxDev** [comment]
You're right, the issue I raised actually cannot happen.
This implies that noop usage is restricted to non-typed containers. A strongly typed array of NO-OP is valid but doesn't make sense does it ?
Since the topic of this thread is noop clarification, here is an other suggestion. I think the website misses an example of the interaction between objects and noop. Noop is described as a valueless value to be added to the _elements_ of a container. The only thing that makes sense is it treat noop as a key-value pair, but one can imagine cases like `key : noop`, so an example would certainly help.

**Comment by tsieprawski** [comment]
@DbxDev Yeah, noop-typed containers are weird. Semantically array is empty (just the length of array-on-the-fly and array-after-ignoring-noops are different), but with objects there go keys+noops (which make no sense).

Maybe noops should not be allowed at all in containers? Or at least in non-optimized containers. Because, noop-typed make no sense (as above), and count-only optimized end up having less items than before parsing. For non-optimized ones, noops still serve as keepalive, and they can stay here.

**Comment by DbxDev** [comment]
> Maybe noops should not be allowed at all in containers?
If so then when can you use them?
Noop in general is not something I find well designed, because it is mixing a definition/action that belongs to a protocol (keep alive) inside a data format. I'm pretty sure that was already discussed in an other thread. That being said, the edge cases mentioned here deserve at least some examples, maybe a "caveats" section, because any parser has to carefully handle them.

**Comment by MikeFair** [comment]
My understanding of noop is for streaming channels and byte alignment issues.

When you have a keep alive requirement in your value stream, you have to send "something" or the connection gets closed on you; send a Noop.  Also if you would like to pad the end of your data stream to align on a better byte boundary for retrieving the next data element, a Noop can be used to fill the space.

[
Also, as a side note related to Noop, there's a proposal floating around to change the Noop value to the "space" character (as it doesn't print anything to the screen and therefore looks like a Noop).  The issue number for it seems missing; however I think it's a brilliant suggestion and deserves attention.
]

**Comment by ColinH** [comment]
IMHO no-op should be completely removed because, again IMHO, it is a violation of the separation-of-concerns principle.

Further, in the presence of no-op, applying the obvious optimisation of pre-allocating containers for arrays and objects when the size, but not the type, is given, can lead to unnecessarily large allocations since, at the extreme, the container might still be empty at the end.

**Comment by Frizlab** [comment]
See #88 and #89, No-op is now simply an invalid type for a container.
